app/services/concerns/xccdf/rule_groups.rb

```diff
@@ -0,0 +1,35 @@
+# frozen_string_literal: true
+
+module Xccdf
+  # Methods related to saving groups
+  module RuleGroups
+    extend ActiveSupport::Concern
+
+    included do
+      def save_rule_groups
+        @rule_groups ||= @op_rule_groups.map do |op_rule_group|
+          ::RuleGroup.from_openscap_parser(op_rule_group, benchmark_id: @benchmark&.id)
+        end
+
+        ::RuleGroup.import!(@rule_groups, ignore: true)
+      end
+
+      def save_rule_group_parents
+        @rule_groups_with_parents ||= @op_rule_groups.map do |op_rule_group|
+          ::RuleGroup.from_openscap_parser(op_rule_group,
+                                           benchmark_id: @benchmark&.id,
+                                           parent_id: rule_group_for(ref_id: op_rule_group.parent_id)&.id)
+        end
+        ::RuleGroup.import!(@rule_groups_with_parents, on_duplicate_key_update: {
+                              conflict_target: %i[ref_id benchmark_id],
+                              columns: :all
+                            })
+      end
```

Comment:
This iteration is doing too much unnecessary work:
* you already have an array of `RuleGroup` models from the previous iteration that you consume via `rule_group_for`
* you're iterating through all the rules, not just the ones that belong to a group
* you're updating `:all` columns upon a conflict, while (based on its name) the method only should save the `parent_id`

Something like this should be way faster, but please do some benchmarks:
```ruby
@rule_groups_with_parents = @op_rule_groups.select(&:parent_id).map do |op_rule_group|
  rg = rule_group_for(op_rule_group.id)
  rg.parent_id = rule_group_for(op_rule_group.parent_id)
  rg
end
```

Also extracting the lines above into a separate method might allow you to move the `import!` into the `save_rule_groups` method which would allow us to get rid of the method below that is anyway named inconsistently.

Reply:
Yeah, I agree with all of this. I'll make these updates!